### PR TITLE
Feature/bca/space explore pagination

### LINKED ISSUE #3693: M11.4 Spaces | Space summary pagination
Currently space summary is limited to a maximum number of summary, the API will elvolve to support pagination (Pending MSC).
Update SDK and UI to support pagination

~2d

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #3693

This PR add supports for the Space Summary API pagination (deprecation of `org.matrix.msc2946/rooms/{roomId}/spaces` in favor of `org.matrix.msc2946/rooms/{roomId}/hierarchy`)

Element Android only uses the hierarchy API with depth 1. 
The pagination is triggered seamlesly when the user reaches the last item of the list (using epoxy visibility  tracker on a bottom loading item). Pagination results are cached and kept in memory as long as the screen is visible.

This PR is also doing a quick cleaning by removing support for auto-join, that has been postponed for a later release

Here is a sample with some low page size:
sum_pagination

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt [hunk 1]
```diff
@@ -162,6 +169,28 @@ class SpaceDirectoryController @Inject constructor(
                     }
                 }
             }
+            if (hierarchySummary?.nextToken != null) {
+                val paginationStatus = data.paginationStatus[currentRootId] ?: Uninitialized
+                if (paginationStatus is Fail) {
+                    errorWithRetryItem {
+                        id("error_${currentRootId}_${hierarchySummary.nextToken}")
+                        text(host.errorFormatter.toHumanReadable(paginationStatus.error))
+                        listener { host.listener?.retry() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt:178]: I think calling `retry()` is not correct here, you should call `host.listener?.loadAdditionalItemsIfNeeded()`

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryViewModel.kt [hunk 2]
```diff
@@ -93,29 +96,44 @@ class SpaceDirectoryViewModel @AssistedInject constructor(
...
-                val knownSummaries = query.second.mapNotNull {
+                val query = session.spaceService().querySpaceChildren(
+                        spaceId,
+                        limit = 10
+                )
+                val knownSummaries = query.children.mapNotNull {
                     session.getRoomSummary(it.childRoomId)
                             ?.takeIf { it.membership == Membership.JOIN } // only take if joined because it will be up to date (synced)
-                }
+                }.distinctBy { it.roomId }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryViewModel.kt:118]: Why adding this call to `distinctBy()`? The server can send duplicates? If it is the case, it's probably better to handle that SDK side.

File: vector/src/main/java/im/vector/app/features/spaces/manage/SpaceManageRoomsController.kt [hunk 3]
```diff
@@ -88,5 +93,29 @@ class SpaceManageRoomsController @Inject constructor(
...
         }
+        val nextToken = roomListAsync.invoke()?.nextToken
+        if (nextToken != null) {
+            // show loading item
+            val paginationStatus = data.paginationStatus
+            if (paginationStatus is Fail) {
+                errorWithRetryItem {
+                    id("error_$nextToken")
+                    text(host.errorFormatter.toHumanReadable(paginationStatus.error))
+                    listener { host.listener?.retry() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/manage/SpaceManageRoomsController.kt:104]: `host.listener?.loadAdditionalItemsIfNeeded()` ?

**Review by bmarty — COMMENTED** [review]
A few remarks from a static review. Also the CI was not happy, can you have a look please?

File: vector/src/main/java/im/vector/app/features/matrixto/SpaceCardRenderer.kt [hunk 4]
```diff
@@ -62,7 +62,7 @@ class SpaceCardRenderer @Inject constructor(
                 inCard.matrixToAccessImage.isVisible = true
                 inCard.matrixToAccessImage.setImageResource(R.drawable.ic_room_private)
             }
-            val memberCount = spaceSummary.otherMemberIds.size
+            val memberCount = spaceSummary.joinedMembersCount?.let { it - 1 } ?: 0
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/matrixto/SpaceCardRenderer.kt:65]: The summay object here is constructed from space summary API, so it does not have all the info as a real summary from DB would have (like the list of user ids of members). So I switched to using joinedMembersCount, but i believe that joinedMemberCount is counting current user unlike otherMemberIds, so decremented to keep same result as before.
I'll check if I can't just use that number indeed

File: vector/src/main/java/im/vector/app/features/spaces/manage/SpaceManageRoomsController.kt (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/manage/SpaceManageRoomsController.kt:104]: ok

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/space/SpaceChildSummaryResponse.kt [hunk 5]
```diff
@@ -18,21 +18,29 @@ package org.matrix.android.sdk.internal.session.space
 
 import com.squareup.moshi.Json
 import com.squareup.moshi.JsonClass
+import org.matrix.android.sdk.api.session.events.model.Event
 
+/**
+ * The fields are the same as those returned by /publicRooms (see spec), with the addition of:
+ *       room_type: the value of the m.type field from the room's m.room.create event, if any.
+ *       children_state: The m.space.child events of the room.
+ *       For each event, only the following fields are included1: type, state_key, content, room_id, sender,
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/space/SpaceChildSummaryResponse.kt:27]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/space/SpaceService.kt [hunk 6]
```diff
@@ -58,10 +59,18 @@ interface SpaceService {
 
     /**
      * Get's information of a space by querying the server
+     * @param suggestedOnly If true, return only child events and rooms where the m.space.child event has suggested: true.
+     * @param limit a client-defined limit to the maximum number of rooms to return per page. Must be a non-negative integer.
+     * @param maxDepth: Optional: The maximum depth in the tree (from the root room) to return.
+     * @param from: Optional. Pagination token given to retrieve the next set of rooms. Note that if a pagination token is provided,
+     * then the parameters given for suggested_only and max_depth must be the same.
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/space/SpaceService.kt:66]: removed comment

**Review by bmarty — APPROVED** [review]
Thanks for the update!

**Pull request merged**